### PR TITLE
feat: implement vite package bundler and fix previous webpack config

### LINKED ISSUE #1131: Feature: webpack 5 dependency updates
### Is your feature request related to a problem? Please describe.

Some of our dependencies could use some major version / breaking changes updates:
- [css-loader](https://www.npmjs.com/package/css-loader) requires some `file-loader` and `url-loader` updates as described in: [6.0.0](https://github.com/webpack-contrib/css-loader/blob/master/CHANGELOG.md#600-2021-07-14)
- [mini-css-extract-plugin](https://www.npmjs.com/package/mini-css-extract-plugin) requires some `webpack@5` updates as described in: [2.0.0](https://github.com/webpack-contrib/mini-css-extract-plugin/blob/master/CHANGELOG.md#200-2021-06-30)
- [postcss-loader](https://www.npmjs.com/package/postcss-loader) requires some `webpack@5` updates as described in: [5.0.0](https://github.com/webpack-contrib/postcss-loader/blob/master/CHANGELOG.md#500-2021-02-02)
- [webpack-manifest-plugin](https://www.npmjs.com/package/webpack-manifest-plugin) requires some `webpack@5` updates as described in: [3.0.0](https://github.com/shellscape/webpack-manifest-plugin/releases/tag/v3.0.0)
- [webpack](https://www.npmjs.com/package/webpack) requires a lot of updates as described in: [5.0.0](https://webpack.js.org/blog/2020-10-10-webpack-5-release/)
- [webpack-dev-server](https://www.npmjs.com/package/webpack-dev-server) requires a few minor updates as described in: [4.0.0](https://github.com/webpack/webpack-dev-server/blob/master/CHANGELOG.md#400-rc0-2021-07-19)

### Describe the solution you'd like

Update webpack to version 5, update all the dependencies, then ping a maintainer to help with the major webpack updates - doing this without the peer dependencies is gruesome if not potentially impossible and should be handled in one go!

### Describe alternatives you've considered

Switch to snowpack, which is a great idea, however, without this update, we cannot accurately tell how good/big of an improvement snowpack is.

### LINKED ISSUE #1346: Feature: investigate vite-plugin-pwa for workbox
### Type of feature

Build

### Current behavior

create-react-app@5 had some interesting workbox updates to produce the manifest and populate it with relevant data from a static config, something we are doing manually right now

### Desired solution

integrate [vite-plugin-pwa](https://github.com/antfu/vite-plugin-pwa) with sensible create-react-app@5 defaults

### Additional context

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by mtfoley**
## What type of PR is this? (check all applicable)

- [x] ♻️ Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [x] 👷 Optimization
- [ ] 📝 Documentation Update
- [ ] 🔖 Release
- [ ] 🚩 Other

## Description

One could argue this closes #1131 😝   I managed to get things running in like 25 minutes because it had so many steps in common with PR #1320 💨 

## Related Tickets & Documents

closes #1346

## Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [x] 🙋 no, because I need help

## Added to documentation?

- [ ] 📜 readme
- [ ] 📜 contributing.md
- [ ] 📓 docs
- [ ] 📕 storybook
- [ ] 🙅 no documentation needed

## [optional] Are there any post-deployment tasks we need to perform?

Pssh probably

## [optional] What gif best describes this PR or how it makes you feel?
sonic

### CODE REVIEW

**Comment by 0-vortex** [comment]
add `closes #` somewhere :D

**Comment by mtfoley** [comment]
Kinda sorta works at this point, but the service worker is broken, and that's breaking a lot of things in turn.  Going to try some of what's in here: https://vite-plugin-pwa.netlify.app/guide/inject-manifest.html#custom-service-worker

**Comment by 0-vortex** [comment]
adjusted the name of the PR to something more descriptive of what you're doing, now you can `closes #` the webpack issue 🍕

**Comment by mtfoley** [comment]
@0-vortex I can reproduce the error locally with `node node_modules/vite/bin/vite.js preview` after running a build. Per your suggestion, I tried this with https://github.com/vitejs/vite/tree/main/packages/plugin-react and it was even more messed up. I'm going to see if I can isolate a problematic component for now.

**Comment by 0-vortex** [comment]
our [.babelrc](https://github.com/open-sauced/open-sauced/blob/main/.babelrc) might have something to do with it as well!

**Comment by mtfoley** [comment]
So far it doesn’t appear to be an issue with the service worker (anymore), and at the moment the error seems to be somewhere in or near the react markdown component we use.

**Comment by 0-vortex** [comment]
> So far it doesn’t appear to be an issue with the service worker (anymore), and at the moment the error seems to be somewhere in or near the react markdown component we use.

Hope this is not a longshot but that honestly sounds much more reasonable, we might get some help around that by searching for the rollup keyword and our failing component :D

**Comment by mtfoley** [comment]
> So far it doesn’t appear to be an issue with the service worker (anymore), and at the moment the error seems to be somewhere in or near the react markdown component we use.

This appears to be most likely related:
https://github.com/vitejs/vite/issues/3592

cc @ajcwebdev since I know you were looking at this also.

**Comment by mtfoley** [comment]
ugh, rock and a hard place:

Up until now I've been trying to leave everything related to testing untouched. 

1) the issue we were running into was in fact related to `react-markdown`. we had been at v6.0.3. There was a downstream dependency that had a circular reference that only showed up in ESM.
2) In order to resolve it, I tried to upgrade to something in v7. The big change for v7 was transitioning to ESM instead of CommonJS.
3) The build worked in the deploy preview worked great, but when I merged upstream changes to main, tests stopped working, and their complaining about something in `react-markdown`. 
3) at this point, Jest has experimental support for ESM, and it requires running node with flags.

I've read a little bit about using vite and jest, but this scope be creepin'!

**Comment by 0-vortex** [comment]
> This appears to be most likely related: [vitejs/vite#3592](https://github.com/vitejs/vite/issues/3592)

If the package is free we could fork it, patch it in our fork, load it in this repo and have the upstream guys look at this in another repository. That's mostly concerning the local fix mentioned in that post :D

**Comment by 0-vortex** [comment]
try rebasing from #1333 got everything I could think of fixed except `webpack@5` and `react-router@6` 🔥

**Comment by mtfoley** [comment]
> try rebasing from #1333 got everything I could think of fixed except `webpack@5` and `react-router@6` 🔥

I actually don't know how to rebase the PR to a different PR, but I think I'd rather either wait for #1333 to land, or I can try doing the `react-markdown` mock that's done in #1333 to see if that resolves it within the scope of this one.

**Comment by mtfoley** [comment]
@0-vortex thanks for all the help upstream on this.  I think the changes as of late had a lot of effects on functionality related to service worker, so I've made enough adjustments for the way vite handles ENV, that it will get as far as resolving to a service-worker.js file. at the moment, it looks like this file came from CRA and replaced the home grown one we had.  The CRA one is written as ESM but it doesn't get transformed so the browser doesn't like it (cannot import outside of module).  Do you think we should update main so that the original service-worker.js file is used?

**Comment by mtfoley** [comment]
@bdougie well this one has been fun, and a lot more involved than I had expected! check out the deploy preview, looks like service worker is loading now.

**Comment by mtfoley** [comment]
@bdougie following up on some changes from @0-vortex yesterday. We had a pretty naive implementation of vite (as evident by small config file).  It had been functional, but hadn't reached feature parity with webpack.

His last PR to my feature branch did a bunch of things, but a handful of notable ones:
1) introduces visual change to include version number in page title
2) reworks serviceworker to use `workbox` rather than homemade file
3) introduces `npm run build:analyze` to create build report (love it!)
image

File: package.json [hunk 1]
```diff
@@ -62,87 +61,47 @@
     "react-router-dom": "^5.2.1",
     "styled-components": "^5.3.3",
     "svgmoji": "^3.2.0",
-    "swr": "^1.1.1",
-    "workbox-background-sync": "^6.4.2",
```
Comment (0-vortex) [inline package.json:66]: @mtfoley we had workbox :p it was just a weird hybrid

File: src/components/Footer.jsx [hunk 2]
```diff
@@ -2,12 +2,12 @@ import React from "react";
...
 import {fontSize} from "../styles/variables";
 
-function Footer() {
+function Footer({date}) {
   return (
     <footer>
       <small style={{fontSize: fontSize.small}}>
         Made with <PrimaryLink href="https://www.onegraph.com/">OneGraph</PrimaryLink>. Hosted on{" "}
-        <PrimaryLink href="https://www.netlify.com/">Netlify</PrimaryLink>.
+        <PrimaryLink href="https://www.netlify.com/">Netlify</PrimaryLink>. Built at {date}
```
Comment (0-vortex) [inline src/components/Footer.jsx:10]: this date could use some formatting, meybe undo the `Open Sauced vX.X.X` title and add this metadata in the <head>

File: src/index.jsx [hunk 3]
```diff
@@ -24,6 +24,22 @@ function Index() {
   const [isAdmin, setIsAdmin] = useState(null);
 
   useEffect(() => {
+    registerSW({
```
Comment (0-vortex) [inline src/index.jsx:27]: this is just a baseline, we would have to decide on what to do here, integrate and update interval, remove logging, extend it?

File: vite.config.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+import ViteReplace from '@rollup/plugin-replace'
+import ViteVisualizer from 'rollup-plugin-visualizer'
+
+export default defineConfig(({command, mode, ...rest }) => {
+  // figure out commands
+  const isBuild = command === 'build';
+
+  // figure out modes
+  const isDev = mode === "development";
+  const isProd = mode === "production";
```
Comment (0-vortex) [inline vite.config.ts:17]: this warning should not break our production build, like it previously would

**Comment by 0-vortex** [comment]
LGTM! 👍 

some minor improvement chores we can open issues for but keen to get this in 2021 still 🍕

**Comment by 0-vortex** [comment]
@bdougie any chance you can check out this new technique?

File: src/index.jsx (hunk 3 above)
Comment (0-vortex) [reply src/index.jsx:27]: have 2-3 issues worth of ideas here, will ping back later

**Pull request merged**